plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts

```diff
@@ -99,106 +100,98 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
         }),
       );
     }
-    this.running = true;
-    this.run();
-  }
 
-  private async run() {
-    while (this.running) {
-      try {
-        // TODO: We want to disconnect the queue popping and message processing
-        // so that if the queue popping fails we exponentially back off in order to give the DB room to sort itself out.
-        await this.process();
-      } catch (e) {
-        this.logger.warn('Processing failed with:', e);
-        // TODO: this can be a little smarter as mentioned in the above comment.
-        // But for now, if something fails, wait a brief time to pick up the next message.
-        await this.wait();
-      }
+    if (this.stopFunc) {
+      throw new Error('Processing engine is already started');
     }
-  }
 
-  private async process() {
-    const { items } = await this.processingDatabase.transaction(async tx => {
-      return this.processingDatabase.getProcessableEntities(tx, {
-        processBatchSize: 1,
-      });
-    });
+    this.stopFunc = startTaskPipeline<RefreshStateItem>({
+      lowWatermark: 5,
+      highWatermark: 10,
+      loadTasks: async count => {
+        try {
+          const { items } = await this.processingDatabase.transaction(
+            async tx => {
+              return this.processingDatabase.getProcessableEntities(tx, {
+                processBatchSize: count,
+              });
+            },
+          );
+          return items;
+        } catch (error) {
+          this.logger.warn('Failed to load processing items', error);
+          return [];
+        }
+      },
+      processTask: async item => {
```

Comment:
The startTaskPipeline call is a bit on the large side :) maybe worth it to make private methods or functions out of these

Reply:
Made me think. Suggested interface tweak:

```ts
const limiter = createTaskLimiter({ lowWatermark, highWatermark });
const stopFn = limiter.run(readBatchFn, processItemFn);
```

Reply:
Had that before, but was just a bunch of boilerplate so got rid of it. If needed the processing function could be extracted but I think this is kinda fine tbh